enginetest/queries/json_scripts.go

```diff
@@ -330,4 +330,102 @@ var JsonScripts = []ScriptTest{
 			},
 		},
 	},
+	// from https://dev.mysql.com/doc/refman/8.0/en/json.html#json-converting-between-types:~:text=information%20and%20examples.-,Comparison%20and%20Ordering%20of%20JSON%20Values,-JSON%20values%20can
+	{
+		Name: "json is ordered correctly",
+		SetUpScript: []string{
+			"create table t (pk int primary key, col1 json);",
+			"insert into t values (1, '{\"items\": {\"1\": 1, \"2\": 2}}');",
+			"insert into t values (2, null);",
+			"insert into t values (3, '{}');",
+			"insert into t values (4, '{\"items\": null}');",
+			"insert into t values (5, (select json_extract('{\"a\": null}', '$.a')));",
+			"insert into t values (6, 0);",
+		},
+		Assertions: []ScriptTestAssertion{
+			{
+				Query: "select * from t order by col1 asc;",
+				Expected: []sql.Row{
+					{2, nil},
+					{5, sql.MustJSON("null")},
+					{6, sql.MustJSON("0")},
+					{3, sql.MustJSON("{}")},
+					{4, sql.MustJSON("{\"items\": null}")},
+					{1, sql.MustJSON("{\"items\": {\"1\": 1, \"2\": 2}}")},
+				},
+			},
+			{
+				Query: "select * from t order by col1 desc;",
+				Expected: []sql.Row{
+					{1, sql.MustJSON("{\"items\": {\"1\": 1, \"2\": 2}}")},
+					{4, sql.MustJSON("{\"items\": null}")},
+					{3, sql.MustJSON("{}")},
+					{6, sql.MustJSON("0")},
+					{5, sql.MustJSON("null")},
+					{2, nil},
+				},
+			},
+		},
+	},
```

Comment:
Wanted to leave this here as a sanity check. Here is the output of InnoDb MySQL:

```
mysql> show create table t;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `pk` int NOT NULL,
  `col1` json DEFAULT NULL,
  PRIMARY KEY (`pk`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select * from t order by pk;
+----+-----------------------------+
| pk | col1                        |
+----+-----------------------------+
|  1 | {"items": {"1": 1, "2": 2}} |
|  2 | NULL                        |
|  3 | {}                          |
|  4 | {"items": null}             |
|  5 | null                        |
|  6 | 0                           |
+----+-----------------------------+
6 rows in set (0.00 sec)

mysql> select * from t order by col1;
+----+-----------------------------+
| pk | col1                        |
+----+-----------------------------+
|  2 | NULL                        |
|  5 | null                        |
|  6 | 0                           |
|  3 | {}                          |
|  4 | {"items": null}             |
|  1 | {"items": {"1": 1, "2": 2}} |
+----+-----------------------------+
6 rows in set, 1 warning (0.00 sec)
```